`Development`: Re-implement integration of feedback suggestion service Athena

pal03377 opened the pull request:
---

# Do not deploy on normal test server
> **Warning**
> This PR includes database changes. Only deploy it on [ma-schwind.ase.cit.tum.de](https://ma-schwind.ase.cit.tum.de/) (see testing steps).

---

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [X] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [X] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [X] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [X] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [X] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [X] I documented the Java code using JavaDoc style.
#### Client
- [X] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [X] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [X] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage) while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [X] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [X] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [X] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Athena is an additional service for Artemis for giving (semi-)automatic feedback suggestions on text exercises.
We recently updated Athena to use a new and much more modular system. This will later enable using Athena for other exercise types and with different approaches to generate suggestions (like LLMs).

The current version of Athena ([ls1intum/Athena-CoFee](https://github.com/ls1intum/Athena-CoFee)) using the "CoFee" approach by @jpbernius is not operational right now. We want to directly connect the CoFee module of the new version through the new Athena architecture from [ls1intum/Athena](https://github.com/ls1intum/Athena) with Artemis.

### Description
<!-- Describe your changes in detail -->
- Removed most old Athena (formerly Athena) code. That's the reason for the large number of deleted lines. In other placed, I renamed "Athene" to "Athena". This PR includes #6862.
  - https://github.com/ls1intum/Artemis/labels/database Removed clusters - they now fully live in the Athena service for less coupling.
  - Removed cluster statistics feature because it needed access to clusters and was rarely used.
  - Moved all Athena-related tasks described below into Athena, notably submission selection and generation of feedback suggestions.
- Added four Athena endpoints, which can be seen below. Also, I adapted the health info provider to work with Athena again.
![Architecture between Artemis and Athena](https://github.com/ls1intum/Artemis/assets/9006596/633aba3e-eef4-4558-a0eb-6e9c9f95ef29)

---

### Deployment

1. Ensure the deployment of the Athena test server:
    - Go to https://github.com/ls1intum/Athena/pull/67
    - Add the `deploy:athena-test3` label like you would on an Artemis PR
    - After a short time (3-6 minutes), verify the deployment in the footer on https://athenetest1-03.ase.cit.tum.de/playground
2. Deploy this PR to the `ma-schwind` test server like you would normally:

https://github.com/ls1intum/Artemis/assets/9006596/3d65aa1e-42ce-4539-918e-09cecf747543

(Alternative to the steps above: Quickly hop on a call with @pal03377 and also let him help you with testing)

---

### Steps for Testing

#### Preparation
- Visit [ma-schwind.ase.cit.tum.de](https://ma-schwind.ase.cit.tum.de/) and log in with [one of the usual](https://confluence.ase.in.tum.de/x/lVGBAQ) **instructor** accounts.
- Choose one of the exercises in the prepared courses called `Athena Test Course`[^1]

#### Submission Sending
Athena sends submissions when the deadline for an exercise is reached (see diagram above).
- Change the deadline of the exercise to be in 1 minute.
- Wait for the deadline.
- Check that Athena has received the submissions at the bottom of [https://athenetest1-03.ase.cit.tum.de/logs](https://athenetest1-03.ase.cit.tum.de/logs). You can find the credentials on [the test servers page in Confluence](https://confluence.ase.in.tum.de/x/lVGBAQ). The endpoint is called `/submissions`. You will probably need to scroll up quite a bit (maybe Strg/Cmd+F for `/submissions` to accelerate this) because the CoFee service itself spits out quite a few messages. There **might** be multiple requests to `/submissions` because the request is batched into chunks of 100 for performance reasons.
 
![expected-submission-sending](https://github.com/ls1intum/Artemis/assets/9006596/293c95a4-7306-4d56-b77e-a8bae3f42327)

> **Note**
> From now on, you can "downgrade" your login to a TA.

#### Submission Selection
Athena tells Artemis which submission is best to assess next.
- <detail><summary>Start to assess a **new** submission.</summary> Navigate to the exercise in the course management. Go to "Exercise Dashboard". If your user hasn't already done this, click the blue button titled "Start participating in the exercise". Click on the green button "Start new assessment".</detail>
- Check the logs like before. There should be a request to get the optimal submission to grade. The endpoint is called `/select_submission`. You might need to scroll up a bit again because the feedback selection endpoint will be called right after the selection endpoint. The processing is finished, and you can continue when a message like `Received 11 clusters and 56 segments from CoFee` appears (for the pre-existing exercises, this might already be done).
- If you are >= 2 people (e.g., in a testing session): Try clicking the "Start assessment" buttons at exactly the same time to provoke problems with getting the same submission to grade.

![expected-submission-selection](https://github.com/ls1intum/Artemis/assets/9006596/1831bd72-df85-4777-a236-3c04f6322298)

**Feedback Suggestions**
Athena provides feedback suggestions in the assessment editor.
- Depending on the submission, you might already get a feedback suggestion.
- Otherwise, give manual feedback on the submission and submit it. Then assess the next one. If it is very similar to an already assessed submission, you should get your own feedback from before as the new suggestion.
- Optional: The Athena endpoint is called `/feedback_suggestions`.

![expected-feedback-suggestions](https://github.com/ls1intum/Artemis/assets/9006596/6880a4ee-d49f-4900-b6fb-e17296ef29b5)

HTTP errors related to the `/text-assessment/events` endpoint are expected: These are also an issue on develop, but they only appear when automatic text assessment is enabled. Fixing this is outside the scope of this PR.

![screenshot-2023-08-05_001752](https://github.com/ls1intum/Artemis/assets/9006596/4b562234-4557-4858-9049-fb92c0a6ddd6)

#### Feedback Sending
Athena receives given feedback when you submit the assessment.
- Save the assessment with feedback. Don't submit it yet.
- Nothing should be sent to Athena: Check the logs like before. The endpoint is called `/feedback`.
- Now, submit the assessment.
- Check the logs like before. Athena should now receive the feedback. It will automatically de-duplicate it by itself.

![expected-feedbacks](https://github.com/ls1intum/Artemis/assets/9006596/99ed97e9-5a1b-4c39-a56a-80da7bcd6c6f)

#### Health
Artemis knows about the health status of the Athena service.
- Log in as an admin (credentials on [the Test Servers Confluence page](https://confluence.ase.in.tum.de/x/lVGBAQ).
- Visit https://ma-schwind.ase.cit.tum.de/admin/health
- View the health of Athena. It should be up. When you view the details, some of the listed modules might be down, which is ok. 
- View the details. The `module_text_cofee` should be running.

![expected-health](https://github.com/ls1intum/Artemis/assets/9006596/4f3ecbb6-b212-4fa3-a147-5a1788a1f989)

#### Exam Mode
<!-- If this PR changes some components that are also used in the exam mode, the PR needs additional testing that the exam mode is still working as expected. -->
<!-- If the testing steps above already describe the exam mode or the exam mode cannot be affected by this PR in any way, you can leave this out. -->
Athena currently does not work in Exam mode. Check that you cannot enable it on text exercises in exams.

#### Breaking Stuff
Try to find edge cases in which the system breaks. Examples: Try opening existing assessments (no new feedback suggestions should appear). Try using the "assess next" button. Temporarily switch off the "Automatic assessment suggestions enabled" checkbox in the exercise settings (and please turn it on again afterward). Set the same due date for both exercises. Be creative! 🛠📋

---

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
- [X] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [X] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

### Test Coverage
#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [athena.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/assessment/athena.service.ts.html) | 93.75% | ✅ |
| text-exercise-cluster-statistics.model.ts | deleted |  |
| exam-management.route.ts | deleted |  |
| [text-submission-assessment.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/text/assess/text-submission-assessment.component.ts.html) | 91.91% | ✅ |
| cluster-statistics.component.ts | deleted |  |
| cluster-statistics.route.ts | deleted |  |
| [text-exercise.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/text/manage/text-exercise/text-exercise.service.ts.html) | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [AthenaAuthorizationInterceptor.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.config.auth/AthenaAuthorizationInterceptor.html) | 100% |  |
| [Feedback.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/Feedback.html) | 97% |  |
| [TextBlock.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/TextBlock.html) | 96% |  |
| [TextBlockRef.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/TextBlockRef.html) | 19% | Record without any methods |
| TextCluster.java | deleted |  |
| [TextExercise.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/TextExercise.html) | 100% |  |
| [NetworkingError.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.exception/NetworkingError.html) | 0% | It's a really basic error class |
| [FeedbackRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/FeedbackRepository.html) | 83% | only removed code |
| TextBlockRepository.java | deleted |  |
| TextClusterRepository.java | deleted |  |
| [TextSubmissionRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/TextSubmissionRepository.html) | 68% | only new database queries |
| AutomaticTextFeedbackService.java | deleted |  |
| [SubmissionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/SubmissionService.html) | 91% |  |
| TextAssessmentQueueService.java | deleted |  |
| [TextAssessmentService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/TextAssessmentService.html) | 100% | ✅ |
| [TextSubmissionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/TextSubmissionService.html) | 92% | ✅ |
| [AthenaFeedbackSendingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.athena/AthenaFeedbackSendingService.html) | 92% | ✅ |
| [AthenaFeedbackSuggestionsService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.athena/AthenaFeedbackSuggestionsService.html) | 90% | ✅ |
| [AthenaHealthIndicator.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.athena/AthenaHealthIndicator.html) | 96% | ✅ |
| [AthenaSubmissionSelectionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.athena/AthenaSubmissionSelectionService.html) | 100% | ✅ |
| [AthenaSubmissionSendingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.athena/AthenaSubmissionSendingService.html) | 93% | ✅ |
| [AthenaConnector.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.athena/AthenaConnector.html) | 84% |  |
| AtheneHealthIndicator.java | deleted |  |
| AtheneService.java | deleted |  |
| InstanceMessageSendService.java | deleted |  |
| [AthenaScheduleService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.scheduled/AthenaScheduleService.html) | 46% | wasn't tested before as well |
| AtheneScheduleService.java | deleted |  |
| [AthenaResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/AthenaResource.html) | 91% | ✅ |
| [TextAssessmentResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/TextAssessmentResource.html) | 90% | ✅ |
| TextClusterResource.java | deleted |  |
| [TextSubmissionResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/TextSubmissionResource.html) | 97% | ✅ |
| AdminTextExerciseResource.java | deleted |  |
| TextClusterStatisticsDTO.java | deleted |  |
| AtheneResource.java | deleted |  |



[^1]: You can also create your own course to test, but you will need at least 10 (!) submissions in it for Athena to suggest feedback. Because this is a bit annoying, I've prepared it for you. If you decide to make your own exercise, make sure that `Automatic assessment suggestions` are enabled and that all submissions are detected to be in English by Artemis (otherwise CoFee will ignore them).

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaFeedbackSendingService.java:
```diff
@@ -0,0 +1,122 @@
+ */
+@Service
+@Profile("athena")
+public class AthenaFeedbackSendingService {
```
What do you think of `AthenaFeedbackPublishingService`?

pal03377 replied:
I equally like that name, but then I would have to refactor Athena a bit as well (which is fine if there's a real advantage in the new name). I'll ask @FelixTJDietrich and @maximiliansoelch about their opinions as well and then come back to this.

FelixTJDietrich replied:
I think there is not much of a difference between both. I left another comment asking about feedback changes and deletions, we probably also have to handle those. In that case, Athena would subscribe to feedback and then the publishing name would make more sense to me.

pal03377 replied:
I briefly changed this but then reverted the change because I think the verb `send` more intuitively captures that the service only performs an HTTP request. Athena does not really subscribe to anything; Artemis actively sends the submissions and the feedback to Athena.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaFeedbackSendingService.java:
```diff
@@ -0,0 +1,122 @@
+        log.debug("Start Athena Feedback Sending Service for Text Exercise '{}' (#{}).", exercise.getTitle(), exercise.getId());
+
+        if (feedbacks.isEmpty()) {
+            log.info("No feedback given for submission #{}.", submission.getId());
```
I don't think this is an actual information.
```suggestion
            log.debug("No feedback given for submission #{}.", submission.getId());
```

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaHealthIndicator.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    private static String moduleHealthToString(AthenaModuleHealth moduleHealth) {
+        var healthString = "";
+        if (moduleHealth.healthy) {
+            healthString += GREEN_CIRCLE; // green circle
+        }
+        else {
+            healthString += RED_CIRCLE; // red circle
+        }
```
```suggestion
        var healthString = moduleHealth.healthy ? GREEN_CIRCLE : RED_CIRCLE;
```

pal03377 replied:
Good improvement 👍

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSelectionService.java:
```diff
@@ -0,0 +1,113 @@
+        }
+    }
+
+    private record ResponseDTO(long data // will contain the submission ID to choose, or -1 if no submission was explicitly chosen
```
```suggestion
    private record ResponseDTO(long submissionId // will contain the submission ID to choose, or -1 if no submission was explicitly chosen
```

pal03377 replied:
Good suggestion! I initially thought I had to use this design because Athena sends a module response wrapped in a JSON object with `data` as the actual contents, but now I discovered `@JsonProperty("data") long submissionId` and use it.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSelectionService.java:
```diff
@@ -0,0 +1,113 @@
+     * @return a Submission suggested by the Athena submission selector (e.g. chosen by the highest information gain)
+     * @throws IllegalArgumentException if exercise isn't automatically assessable
+     */
+    public Optional<TextSubmission> getProposedSubmission(TextExercise exercise, List<TextSubmission> submissions) {
```
Do you really need all submissions (objects) here? I think it would be better if you just use the ids here and then load only the one selected submission from the database. Keep in mind that we have courses with up to 2000 students, so a list of 2000 text submission would be needed here.

pal03377 replied:
It makes sense to split it up 👍  Somehow, the old Athena worked despite always sending all submissions at once, but this solution is better - so I will change that.

pal03377 replied:
Good input! I changed this function to get submission IDs. However, I kept the submission objects in `TextSubmissionService.getRandomTextSubmissionEligibleForNewAssessment`. There are two reasons for this:
1. The logic used in `SubmissionService.getAssessableSubmissions` is quite complex and hard to get in a pure database call, so I'm keeping it as-is.
2. With my latest changes in https://github.com/ls1intum/Artemis/pull/6861/commits/9c33e20d6dda53b8b958504c01dd5018444e929d, the memory usage should be as before (loading the submissions exactly once). To improve this, one would have to refactor quite a lot (e.g., using streams only instead of lists?). I don't want to do that in this PR, if possible, as it's quite large by itself already.

pal03377 replied:
Good input! I changed this function to get submission IDs. However, I kept the submission objects in `TextSubmissionService.getRandomTextSubmissionEligibleForNewAssessment`. There are two reasons for this:
1. The logic used in `SubmissionService.getAssessableSubmissions` is quite complex and hard to get in a pure database call, so I'm keeping it as-is.
2. With my latest changes in https://github.com/ls1intum/Artemis/pull/6861/commits/9c33e20d6dda53b8b958504c01dd5018444e929d, the memory usage should be as before (loading the submissions exactly once). To improve this, one would have to refactor quite a lot (e.g., using streams only instead of lists?). I don't want to do that in this PR, if possible, as it's quite large by itself already.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSelectionService.java:
```diff
@@ -0,0 +1,113 @@
+            if (submissionId == -1) {
+                return Optional.empty();
+            }
+            return submissions.stream().filter(s -> s.getId() == submissionId).findFirst();
```
Please check (maybe not here but after calling the method) if the selected submission is still assessable (i.e. did not get locked in the meantime by a different user). I assume that getting the best available submission is deterministic, meaning if two tutors press "assess" at the same time, they would try to both assess the same submission.

pal03377 replied:
It will depend on the module, but checking is important regardless 👍 I've added that.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSendingService.java:
```diff
@@ -0,0 +1,104 @@
+        log.debug("Start Athena Submission Sending Service for Text Exercise '{}' (#{}).", exercise.getTitle(), exercise.getId());
+
+        // Find all text submissions for exercise (later we will support others)
+        Set<TextSubmission> textSubmissions = textSubmissionRepository.getTextSubmissionsWithTextBlocksByExerciseId(exercise.getId());
```
Split this up into sub-requests and use Pageables to retrive submissions, we should not try to deal with 2000 object in memory at the same time.

pal03377 replied:
It makes sense to split it up 👍 Somehow, the old Athena worked despite always sending all submissions at once, but this solution is better - so I will change that.

pal03377 replied:
Added batching! 🎉

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/dto/athena/TextExerciseDTO.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * A DTO representing a TextExercise, for transferring data to Athena
+ */
+public record TextExerciseDTO(long id, String title, Double maxPoints, double bonusPoints, String gradingInstructions, String problemStatement) {
```
Why does Athena need the GradingInstructions?

FelixTJDietrich replied:
Athena CoFee does not require them, but in my modules I'm using the grading instructions for the LLM prompt.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/dto/athena/TextFeedbackDTO.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * A DTO representing a Feedback, for transferring data to Athena
+ */
+public record TextFeedbackDTO(long id, long exerciseId, long submissionId, String text, String detailText, double credits, Integer indexStart, Integer indexEnd) {
```
This could lead to issues when tutors use structured grading instructions. Here the feedback is not only the detail text, but also created by the grading instruction. If Athena decides to apply such a feedback to another submission, the grading instruction context gets lost.

pal03377 replied:
I've added a new `grading_instruction` field for feedback in Athena as that seemed like the most straightforward solution to the problem.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/scheduled/AthenaScheduleService.java:
```diff
@@ -0,0 +1,111 @@
+            // NOTE: if you want to test this locally, please comment it out, but do not commit the changes
+            return;
+        }
+        final List<TextExercise> runningTextExercises = textExerciseRepository.findAllAutomaticAssessmentTextExercisesWithFutureDueDate();
```
Since this is the only place that uses this method, we should probably filter for `isFeedbackSuggestionsEnabled` already in the database to avoid loading to many unrelated exercises.

Please have a look at the production instance and check how many exercises will typically be returned by such a call and if it makes sense to introduce paging here.

pal03377 replied:
This already happens: The DB call fetches all exercises with assessment type `SEMI_AUTOMATIC`. Currently, that's exactly the condition that `isFeedbackSuggestionsEnabled` checks. This will change in the future (we will add an `exercise.feedback_suggestions` column in the DB), but it's just the assessment type for this PR.

However, I now noticed that the `exercise.feedbackSuggestionsEnabled` field is not needed on the client, so I've `@JsonIgnore`-d it now.

pal03377 replied:
The highest number of simultaneous `SEMI_AUTOMATIC` text exercises at once with `release_date < current_date < due_date` was in May of 2020: 4 exercises. I don't think we need paging 🎉 For reference, I used this script to find out: https://gist.github.com/pal03377/7f5eafa6c4e2900955e61713abf038fa

Strohgelaender reviewed (commented):
Had a look at the server code only, here are a few sugesstions and some potentailly missed special cases:

pal03377 commented on src/main/java/de/tum/in/www1/artemis/domain/TextBlockRef.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ * A TextBlockRef. It is a combining class for TextBlock and Feedback.
+ */
+public class TextBlockRef implements Serializable {
```
Great idea! Saves 50+ lines 💪

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSelectionService.java:
```diff
@@ -60,46 +63,52 @@ private record ResponseDTO(@JsonProperty("data") long submissionId // submission
+                log.error("Athena returned a submission ID that does not exist in the database: {}", response.submissionId);
+                return Optional.empty();
+            }
+            return submission.map(s -> (TextSubmission) s);
```
This returns the submission to later load the submission with results again. You can also just return the ID here and let the caller load the submission.

pal03377 replied:
Good catch, did that 👍

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/TextSubmissionService.java:
```diff
@@ -139,11 +125,24 @@ public Optional<TextSubmission> getRandomTextSubmissionEligibleForNewAssessment(
+            if (athenaSubmission.isPresent()) {
+                // test again if it is still assessable (Athena might have taken some time to respond and another assessment might have started in the meantime)
+                var submission = textSubmissionRepository.findWithEagerResultsAndFeedbackAndTextBlocksById(athenaSubmission.get().getId());
+                if (submission.isPresent() && submission.get().getLatestResult() == null) { // submission assessable?
```
Is this correct? I'd assume that Athena creates an automatic result that contains the feedback suggestions. 
```suggestion
                if (submission.isPresent() && (submission.get().getLatestResult() == null || !submission..get().getLatestResult().isManual())) { // submission assessable?
```

pal03377 replied:
I've changed that for more robustness 👍 I thought about changing the `!...isManual()` check to `...isAutomatic()`, but I think I also want to ignore existing semi-automatic results, so this should be correct.

FelixTJDietrich commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaFeedbackSendingService.java:
```diff
@@ -0,0 +1,122 @@
+        try {
+            final RequestDTO request = new RequestDTO(exercise, submission, feedbacks, textBlockRepository);
+            // TODO: make module selection dynamic (based on exercise)
+            ResponseDTO response = connector.invokeWithRetry(athenaUrl + "/modules/text/module_text_cofee/feedbacks", request, maxRetries);
```
So what if the sending of feedback fails? Should we keep book about this?

pal03377 replied:
I discussed this with @maximiliansoelch. It doesn't hurt if Athena doesn't get some feedback. The worst that can happen is that the suggestions are slightly more inaccurate or not present. So we don't need to introduce a complex storage-and-resending mechanic here.

FelixTJDietrich commented on src/main/java/de/tum/in/www1/artemis/web/rest/TextAssessmentResource.java:
```diff
@@ -147,9 +153,12 @@
-            saveTextBlocks(textAssessment.getTextBlocks(), textSubmission, textExerciseRepository.findByIdElseThrow(exerciseId), response.getBody().getFeedbacks());
+            final var feedbacksWithIds = response.getBody().getFeedbacks();
+            saveTextBlocks(textAssessment.getTextBlocks(), textSubmission, exercise, feedbacksWithIds);
+            sendFeedbackToAthena(exercise, textSubmission, feedbacksWithIds);
```
What about changes to the feedback? For example, I give a bad review, press save, and then change/delete feedback again. How would we handle this? Athena would need the feedback removed/changed.

pal03377 replied:
I discussed this with @maximiliansoelch. Athena already gets changed feedback (it should be overridden in the Athena DB because it has the same ID). We decided that feedback deletion happens too rarely to be relevant for now, so it should be fine like this for now.

FelixTJDietrich commented on docker/artemis/config/prod.env:
```diff
@@ -1,7 +1,7 @@
 
 # Secret Overrides
-ARTEMIS_ATHENE_BASE64SECRET="YWVuaXF1YWRpNWNlaXJpNmFlbTZkb283dXphaVF1b29oM3J1MWNoYWlyNHRoZWUzb2huZ2FpM211bGVlM0VpcAo="
+ARTEMIS_ATHENA_BASE64SECRET="YWVuaXF1YWRpNWNlaXJpNmFlbTZkb283dXphaVF1b29oM3J1MWNoYWlyNHRoZWUzb2huZ2FpM211bGVlM0VpcAo="
```
Does this need changing? I think you renamed it.

pal03377 replied:
You are right 👍

FelixTJDietrich reviewed (changes requested):
I went though all the code and it mostly looks good :) Great work! 🎉
I left some comments on what I think needs changing.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/apollon/ApollonHealthIndicator.java:
```diff
@@ -29,7 +29,7 @@ public ApollonHealthIndicator(@Qualifier("shortTimeoutApollonRestTemplate") Rest
 
     /**
-     * Ping Athene at /queueStatus and check if the service is available.
+     * Ping Athena at /queueStatus and check if the service is available.
```
You are right. I did this correctly in #6862, but forgot it here.

pal03377 commented on docs/dev/setup.rst:
```diff
@@ -600,36 +599,35 @@ instead of the TUM defaults:
 --------------
 
-The semi-automatic text assessment relies on the Athene_ service.
+The semi-automatic text assessment relies on the Athena_ service.
```
I suggest that we do this in another PR when we actually generalize it in the code as well, but good point 👍

terlan98 reviewed (dismissed):
Client code LGTM 👍 Just left a small question

Also attempted to do a manual test but there were problems with the Athena server, which Paul is currently investigating. I'll test once the issue is fixed.

pal03377 commented on src/main/webapp/app/exercises/text/assess/text-assessment-area/text-assessment-area.component.html:
```diff
@@ -2,6 +2,8 @@
     <jhi-textblock-assessment-card
         *ngFor="let ref of textBlockRefs; let i = index"
         id="text-feedback-block-{{ i }}"
+        [attr.startIndex]="ref.block!.startIndex!"
+        [attr.endIndex]="ref.block!.endIndex!"
```
I think I only used these for debugging purposes anyways, so I've now removed them.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/dto/athena/TextExerciseDTO.java:
```diff
@@ -0,0 +1,28 @@
+     * Create a new TextExerciseDTO from a TextExercise
+     */
+    public static TextExerciseDTO of(@NotNull TextExercise exercise) {
+        return new TextExerciseDTO(exercise.getId(), exercise.getTitle(), exercise.getMaxPoints(), exercise.getBonusPoints(), exercise.getGradingInstructions(),
```
As this is not needed for the CoFee module (for which the grading instruction ID suffices) and is outside the scope of this PR, let's discuss this in a follow-up.

pal03377 commented on docs/dev/setup.rst:
```diff
@@ -600,36 +599,35 @@ instead of the TUM defaults:
 
-The Athene service is running on a dedicated machine and is addressed via
+The Athena service is running on a dedicated machine and is addressed via
 HTTP. We need to extend the configuration in the file
```
I've linked to the Athena documentation now, where this will be explained in the future.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaFeedbackSendingService.java:
```diff
@@ -0,0 +1,122 @@
+
+        try {
+            final RequestDTO request = new RequestDTO(exercise, submission, feedbacks, textBlockRepository);
+            // TODO: make module selection dynamic (based on exercise)
```
I discussed this with @maximiliansoelch. This endpoint should be fine here because we only want to make Athena-CoFee work in this PR.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/dto/athena/TextFeedbackDTO.java:
```diff
@@ -24,18 +26,23 @@ public record TextFeedbackDTO(long id, long exerciseId, long submissionId, Strin
+        feedback.setDetailText(description());
         feedback.setCredits(credits());
+        // The given grading instruction should match the one of this DTO:
+        assert gradingInstructionId() == null || gradingInstructionId().equals(gradingInstruction.getId());
```
We should not use assert in production code. Java will just spike these instructions if `-ea` is not set, which is typically not the case in production environments.

Strohgelaender commented on src/test/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSendingServiceTest.java:
```diff
@@ -1,60 +1,99 @@
 
 class AthenaSubmissionSendingServiceTest extends AthenaTest {
 
+    private final static Language DEFAULT_SUBMISSION_LANGUAGE = Language.ENGLISH;
+
+    private final static String DEFAULT_SUBMISSION_TEXT = "This is a test submission.";
```
```suggestion
    private static final Language DEFAULT_SUBMISSION_LANGUAGE = Language.ENGLISH;

    private static final String DEFAULT_SUBMISSION_TEXT = "This is a test submission.";
```

tobias-lippert commented on src/main/java/de/tum/in/www1/artemis/service/scheduled/AthenaScheduleService.java:
```diff
@@ -0,0 +1,111 @@
+    @PostConstruct
+    private void scheduleRunningExercisesOnStartup() {
+        final Collection<String> activeProfiles = Arrays.asList(env.getActiveProfiles());
+        if (activeProfiles.contains(JHipsterConstants.SPRING_PROFILE_DEVELOPMENT)) {
```
just an optional alternative: We already have a ```ProfilService``` that contains a method ```isDev```. In  my opinion that's a bit cleaner.

pal03377 replied:
That's much better, nice! I didn't know about the `ProfileService` before.

tobias-lippert commented on src/main/java/de/tum/in/www1/artemis/web/rest/AthenaResource.java:
```diff
@@ -0,0 +1,76 @@
+ * REST controller for Athena feedback suggestions.
+ */
+@RestController
+@RequestMapping("/api/athena")
```
According to our server guidelines, routes should not start with a slash.
``` suggestion
@RequestMapping("api/athena/")

tobias-lippert commented on src/main/java/de/tum/in/www1/artemis/web/rest/AthenaResource.java:
```diff
@@ -0,0 +1,76 @@
+     */
+    @GetMapping("exercises/{exerciseId}/submissions/{submissionId}/feedback-suggestions")
+    @EnforceAtLeastTutor
+    public ResponseEntity<List<TextBlockRef>> getFeedbackSuggestions(@PathVariable Long exerciseId, @PathVariable Long submissionId) {
```
personal taste: I prefer @PathVariables to be primitives because if the value is indeed null, this endpoint won't be reached.

pal03377 replied:
I fully agree! Changed

tobias-lippert commented on src/main/resources/config/liquibase/changelog/20230713171500_changelog.xml:
```diff
@@ -0,0 +1,11 @@
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.18.xsd">
+    <changeSet author="pal03377" id="20230713171500-3">
```
almost all our changelogs, do not have the -number in the end. for the id. Is this needed here?

pal03377 replied:
This was an artifact from when I generated the changelog. I've changed it 👍

tobias-lippert commented on src/main/webapp/app/assessment/athena.service.ts:
```diff
@@ -0,0 +1,31 @@
+export class AthenaService {
+    public resourceUrl = 'api/athena';
+
+    constructor(protected http: HttpClient, private profileService: ProfileService) {}
```
codacy

pal03377 replied:
ESLint doesn't let me do this: The pre-commit hook always changes it back :/

tobias-lippert commented on src/main/webapp/app/exam/manage/exam-management.route.ts:
```diff
@@ -787,7 +783,7 @@ export const examManagementRoute: Routes = [
         data: {
             authorities: [Authority.ADMIN, Authority.INSTRUCTOR, Authority.EDITOR],
-            pageTitle: 'artemisApp.exampleSubmission.home.title',
+            pageTitle: 'artemisApp.exampleSubmission.home.LtiServicetitle',
```
could you explain this renaming pls.

pal03377 replied:
I believe I accidentally pasted from my clipboard and didn't notice it when I committed 🤦 I've undone this. Thanks for noticing this!

tobias-lippert commented on src/test/java/de/tum/in/www1/artemis/assessment/ExampleSubmissionIntegrationTest.java:
```diff
@@ -333,8 +333,6 @@ void importExampleSubmissionWithTextSubmission() throws Exception {
 
         TextBlock textBlock = new TextBlock();
-        textBlock.setCluster(null);
-        textBlock.setAddedDistance(0);
```
I understand why we remove the setCluster, but why do we also remove the setAddedDistance?

pal03377 replied:
The `addedDistance` of a text block is related to its cluster and is now also stored in Athena directly. It is used for calculating the next suggested submission, which is now a task of the Athena service -> Artemis does not need this any more.

tobias-lippert commented on src/test/java/de/tum/in/www1/artemis/connector/AthenaRequestMockProvider.java:
```diff
@@ -0,0 +1,190 @@
+    private MockRestServiceServer mockServerShortTimeout;
+
+    @Value("${artemis.athena.url}")
+    private String athenaUrl;
```
can be final, see my remarks in ```AthenaFeedbackSuggestionsService ```

pal03377 replied:
Let's discuss all instances of this here: https://github.com/ls1intum/Artemis/pull/6861#discussion_r1277559597

tobias-lippert commented on src/test/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaFeedbackSuggestionsServiceTest.java:
```diff
@@ -0,0 +1,44 @@
+                jsonPath("$.exercise.title").value(textExercise.getTitle()), jsonPath("$.submission.id").value(textSubmission.getId()),
+                jsonPath("$.submission.text").value(textSubmission.getText()));
+        List<TextBlockRef> suggestions = athenaFeedbackSuggestionsService.getFeedbackSuggestions(textExercise, textSubmission);
+        assertThat(suggestions.get(0).feedback().getText()).isEqualTo("Not so good");
```
will suggestions always only contain one element? Otherwise, get(0) might lead to flakiness.

tobias-lippert commented on src/test/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaHealthIndicatorTest.java:
```diff
@@ -0,0 +1,41 @@
+        athenaRequestMockProvider.mockHealthStatusSuccess(true);
+        final Health health = athenaHealthIndicator.health();
+        assertThat(health.getStatus()).isEqualTo(Status.UP);
+        assertThat(health.getDetails().get("module_example").toString()).contains(GREEN_CIRCLE);
```
suggestion: make ``` "module_example"``` a constant

pal03377 replied:
Did just that 👍

tobias-lippert commented on src/test/java/de/tum/in/www1/artemis/assessment/TutorEffortIntegrationTest.java:
```diff
@@ -69,7 +69,7 @@ void initTestCase() {
         course = courseUtilService.createCourseWithTextExerciseAndTutor(TEST_PREFIX + "tutor1");
         exercise = course.getExercises().iterator().next();
         studentParticipation = studentParticipationRepository.findByExerciseId(exercise.getId()).stream().iterator().next();
```
will there never be more than one participation for the given exercise?

pal03377 replied:
No: `createCourseWithTextExerciseAndTutor` only ever creates one.

tobias-lippert reviewed (commented):
Looks good in general, just some comments and questions.

pal03377 commented on .idea/runConfigurations/Artemis__Server__LocalVC___LocalCI__Athena_.xml:
```diff
@@ -1,13 +1,12 @@
 <component name="ProjectRunConfigurationManager">
-  <configuration default="false" name="Artemis (Server, Athene)" type="SpringBootApplicationConfigurationType" factoryName="Spring Boot">
+  <configuration default="false" name="Artemis (Server, LocalVC &amp; LocalCI, Athena)" type="SpringBootApplicationConfigurationType" factoryName="Spring Boot">
+    <option name="ACTIVE_PROFILES" value="dev,bamboo,bitbucket,jira,artemis,scheduling,local,athena" />
```
Very good point! I've changed the order such that `local` is last 👍

pal03377 commented on docker/artemis/config/prod.env:
```diff
@@ -1,7 +1,6 @@
+SPRING_PROFILES_ACTIVE="artemis,scheduling,athena,prod,docker"
 
 # Secret Overrides
-ARTEMIS_ATHENE_BASE64SECRET="YWVuaXF1YWRpNWNlaXJpNmFlbTZkb283dXphaVF1b29oM3J1MWNoYWlyNHRoZWUzb2huZ2FpM211bGVlM0VpcAo="
```
To be honest, I have yet to learn about the specifics of this env file. I just noticed that the test server works without this, so I removed it. I'll ask one of the deployment experts about this and follow-up here.

pal03377 replied:
I asked Ludwig about this and he showed me this file where it is used: https://github.com/ls1intum/Artemis/blob/develop/docker/artemis.yml

Here is the documentation for the related Spring configuration patterns: https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config

Apparently, this is just a kind of safeguard to ensure that secrets are overwritten. I've re-added the configuration line.

pal03377 commented on docs/dev/setup.rst:
```diff
@@ -221,10 +221,9 @@ You can override the following configuration options in this file.
-       athene:
-           url: http://localhost
-           base64-secret: YWVuaXF1YWRpNWNlaXJpNmFlbTZkb283dXphaVF1b29oM3J1MWNoYWlyNHRoZWUzb2huZ2FpM211bGVlM0VpcAo=
-           token-validity-in-seconds: 10800
```
I don't see the purpose this served initially. When looking at the old Athena code, there was a very basic header secret string check (which is the case now as well): https://github.com/ls1intum/Athena-CoFee/blob/7b1b9b48bd6934215d77d963b5d9c1e4b1f813ed/load-balancer/src/main.py#L43C50-L43C50

So the authentication mechanism had been a simple shared secret all the time. Therefore, I believe the `token-validity-in-seconds` parameter wasn't ever used and can be removed.

jpbernius replied:
I think this was related to the tracking feature we had for research purposes.

terlan98 reviewed (commented):
Tested on TS5 and found some issues 😬 You can find my observations and some questions below:

#### Submission Sending
Seems to work fine, but for some reason, there are 2 log messages for the `/submissions` endpoint. Is it intended, or is the request being sent twice?
<img width="900" alt="Screenshot 2023-07-31 at 21 48 20" src="https://github.com/ls1intum/Artemis/assets/22798314/961495e9-78a5-4e2b-a73b-03012607362d">

#### ⚠️ Submission Selection
When I attempted to start a new assessment for the 'What is Binary Search?' exercise, I got a 500 error. Here are some details that might be helpful: 
`"500 Internal Server Error: \"{\"module_name\":\"module_text_cofee\",\"status\":500,\"data\":\"Internal Server Error\",\"meta\":null}\""`

My 2nd attempt with 'SQL vs NOSQL' exercise did not produce the aforementioned error. The `/select_submission` endpoint was called, but the PR description says there should be a request for`/submission_selection`. Is the endpoint name outdated in the description?
<img width="900" alt="Screenshot 2023-07-31 at 22 03 54" src="https://github.com/ls1intum/Artemis/assets/22798314/b2b63cca-4404-4b60-846e-be2dd44166e5">

Also, I did not see a message like `Received 11 clusters and 56 segments from CoFee` 🙁 

#### ✅ Feedback Suggestions
A POST request was sent to `/feedback_suggestions` and I saw one automatic feedback on the submission.

#### ⚠️ Feedback Sending
Athena sent a POST request to `/feedbacks` when I *saved* the assessment 🙁 

When I submitted the assessment, the same request was sent again. I also got this error on the console:
<img width="600" alt="Screenshot 2023-07-31 at 22 16 15" src="https://github.com/ls1intum/Artemis/assets/22798314/3cb0de9c-03d7-40f4-99e4-1969b2851e29">

#### ✅ Health
Works as expected 👍 

#### Other tests to break stuff
- ✅ 'Assess Next' button 
- ✅ Opening finished assessments
  - No new feedback suggestions appear
- ❌ Deleting a suggested feedback, and trying to add a manual feedback for the same text block  
  - The text block is not clickable after deleting the suggested feedback
  - Selecting the block while pressing alt/option button works, though
- ⚠️  Making changes to the feedback text of a suggested feedback
  - It works but I get this error on the console: <img width="600" alt="Screenshot 2023-07-31 at 22 11 14" src="https://github.com/ls1intum/Artemis/assets/22798314/55a46e62-77d5-497d-afba-345e11b73611">

pal03377 commented on src/main/java/de/tum/in/www1/artemis/repository/GradingInstructionRepository.java:
```diff
@@ -13,6 +11,6 @@
 @Repository
 public interface GradingInstructionRepository extends JpaRepository<GradingInstruction, Long> {
 
-    List<GradingInstruction> findByGradingCriterionId(long gradingCriterionId);
```
I noticed that this method was unused when adding a new one here and removed it. Its removal here does have nothing to do with the actual refactoring.

Generally, a grading criterion differs from a grading instruction in that the instruction has a specified number of credits attached to it and can directly be used as feedback. Therefore it is relevant for Athena. Grading criteria are "only" text hints for grading AFAIK.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/TextAssessmentService.java:
```diff
@@ -46,17 +42,11 @@ public TextAssessmentService(UserRepository userRepository, ComplaintResponseSer
         final Participation participation = textSubmission.getParticipation();
-        final TextExercise exercise = (TextExercise) participation.getExercise();
-
-        final boolean computeFeedbackSuggestions = automaticTextFeedbackService.isPresent() && exercise.isAutomaticAssessmentEnabled();
```
Feedback suggestions are now fetched asynchronously after a tutor has opened the submission. Before that, they were directly created before the submission could be opened. I changed this behavior because of performance concerns (waiting for an additional network request). In case it was particularly slow, I prevent the suggestions from appearing if the tutor adds feedback before the suggestions are incoming (see client code).

pal03377 replied:
Therefore, the `prepareSubmissionForAssessment` method does not need to know whether to compute feedback suggestions

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/TextSubmissionService.java:
```diff
@@ -27,21 +29,21 @@ public class TextSubmissionService extends SubmissionService {
     private final TextSubmissionRepository textSubmissionRepository;
 
-    private final Optional<TextAssessmentQueueService> textAssessmentQueueService;
+    private final Optional<AthenaSubmissionSelectionService> athenaSubmissionSelectionService;
```
You are right; it is! 👍  I found that name to be more fitting, considering that this functionality is pretty tightly coupled to the Athena service.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/TextSubmissionService.java:
```diff
@@ -139,11 +128,24 @@ public Optional<TextSubmission> getRandomTextSubmissionEligibleForNewAssessment(
+            if (athenaSubmissionId.isPresent()) {
+                // test again if it is still assessable (Athena might have taken some time to respond and another assessment might have started in the meantime)
+                var submission = textSubmissionRepository.findWithEagerResultsAndFeedbackAndTextBlocksById(athenaSubmissionId.get());
+                if (submission.isPresent() && (submission.get().getLatestResult() == null || !submission.get().getLatestResult().isManual())) { // submission assessable?
```
Good catch! I noticed that the comment essentially only repeats the one from above, so I moved that one down now.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaConnector.java:
```diff
@@ -1,45 +1,49 @@
 
 /**
- * This class allows for an easy implementation of Connectors to Remote Artemis Services (e.g. the Text Clustering System).
+ * Connector to Athena, a remote Artemis service.
```
I like that idea! I added a small description as an overview.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaConnector.java:
```diff
@@ -53,29 +57,28 @@ private ResponseType invoke(@NotNull String url, @NotNull RequestType requestObj
         }
 
         final ResponseType responseBody = response.getBody();
-        assert responseBody != null;
```
To be honest, I didn't even remember removing that assertion 🙈 Now, I added an additional error check there in case something goes wrong.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/exception/NetworkingError.java:
```diff
@@ -1,8 +1,15 @@
+/**
+ * This exception is thrown when a networking error occurs while communicating with external services.
+ */
 public class NetworkingError extends Exception {
```
Good point. I've renamed it to `NetworkingException` (`NetworkingErrorException` sounded a bit too "double").

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaFeedbackSuggestionsService.java:
```diff
@@ -0,0 +1,103 @@
+    private final Logger log = LoggerFactory.getLogger(AthenaFeedbackSuggestionsService.class);
+
+    @Value("${artemis.athena.url}")
+    private String athenaUrl;
```
I would prefer to leave this as-is because most configuration value injections in Artemis are handled like this. What do you think?

tobias-lippert replied:
While I favor constructor injection and prefer making fields final, I see your point. To be honest, I think we just use field injection in most places because initially it might not have been supported or people didn't know this was possible and others just adapted to what was already there. So, if we want to use this, we should probably open a separate PR that unifies this for all occurrences.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaFeedbackSuggestionsService.java:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * Creates a new AthenaFeedbackSuggestionsService to receive feedback suggestions from the Athena service.
+     */
+    public AthenaFeedbackSuggestionsService(@Qualifier("athenaRestTemplate") RestTemplate athenaRestTemplate, GradingInstructionRepository gradingInstructionRepository) {
```
Thanks for the suggestion, see discussion above

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaHealthIndicator.java:
```diff
@@ -0,0 +1,92 @@
+
+    private final RestTemplate shortTimeoutRestTemplate;
+
+    @Value("${artemis.athena.url}")
```
Let's discuss all instances of this here: https://github.com/ls1intum/Artemis/pull/6861#discussion_r1277559597

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaHealthIndicator.java:
```diff
@@ -0,0 +1,92 @@
+
+    private static String moduleHealthToString(AthenaModuleHealth moduleHealth) {
+        var healthString = moduleHealth.healthy ? GREEN_CIRCLE : RED_CIRCLE;
+        healthString += " " + moduleHealth.url + " (" + moduleHealth.type + ")";
```
I feel like this is more readable and the performance gains are negligible (because this is not within a loop). Besides, the Java compiler performs such optimizations by itself anyways, apparently: https://stackoverflow.com/a/11942386/4306257

Strohgelaender replied:
I agree, outside of loops StringBuilders are not needed, Java optimizes that internally.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaHealthIndicator.java:
```diff
@@ -0,0 +1,92 @@
+        this.shortTimeoutRestTemplate = shortTimeoutRestTemplate;
+    }
+
+    private record AthenaModuleHealth(String type, boolean healthy, String url) {
```
It's about the exercise type (text, programming, modeling, ...), which is relevant because the different modules in Athena are only compatible with one exercise type each. However, I had to think a bit about this myself as well and now decided to rename this variable to `exerciseType` for clarity.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaHealthIndicator.java:
```diff
@@ -0,0 +1,92 @@
+        }
+        catch (Exception emAll) {
+            health = new ConnectorHealth(emAll);
+            health.setAdditionalInfo(Map.of("url", athenaUrl));
```
Personally, I feel like that's a bit overkill here. This way, it is immediately clear what is happening. Also, this value will probably not be changed in the future, so I believe it is valuable to directly see it. What do you think?

pal03377 replied:
Did that! 👍

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaHealthIndicator.java:
```diff
@@ -0,0 +1,92 @@
+     */
+    private Map<String, Object> getAdditionalInfoForModules(JsonNode athenaHealthResponse) {
+        var additionalModuleInfo = new HashMap<String, Object>();
+        if (athenaHealthResponse.has("modules")) {
```
I replaced all JSON keys with `static final` constants now.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSelectionService.java:
```diff
@@ -0,0 +1,114 @@
+
+    private final Logger log = LoggerFactory.getLogger(AthenaSubmissionSelectionService.class);
+
+    @Value("${artemis.athena.url}")
```
Let's discuss all instances of this here: https://github.com/ls1intum/Artemis/pull/6861#discussion_r1277559597

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSelectionService.java:
```diff
@@ -0,0 +1,114 @@
+     * @return a submission ID suggested by the Athena submission selector (e.g. chosen by the highest information gain)
+     * @throws IllegalArgumentException if exercise isn't automatically assessable
+     */
+    public Optional<Long> getProposedSubmission(TextExercise exercise, List<Long> submissionIds) {
```
Good idea! I probably missed this when I changed the behavior of this method to improve performance.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSendingService.java:
```diff
@@ -0,0 +1,133 @@
+    private final Logger log = LoggerFactory.getLogger(AthenaSubmissionSendingService.class);
+
+    @Value("${artemis.athena.url}")
+    private String athenaUrl;
```
Let's discuss all instances of this here: https://github.com/ls1intum/Artemis/pull/6861#discussion_r1277559597

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaFeedbackSuggestionsService.java:
```diff
@@ -0,0 +1,103 @@
+
+    private final Logger log = LoggerFactory.getLogger(AthenaFeedbackSuggestionsService.class);
+
+    @Value("${artemis.athena.url}")
```
Let's discuss all instances of this here: https://github.com/ls1intum/Artemis/pull/6861#discussion_r1277559597

pal03377 commented:
@terlan98 Thanks so much for the detailed review!

- **Submission Sending**: This is not intended, but I also cannot reproduce it 🤔 Could it be that you and I tested the submission sending at the same time accidentally?
- **Submission Selection**: Good catch! I've fixed the error and ensured that a server error from Athena does not affect the ability to assess submissions in the future. Re: Endpoint: You are correct; the endpoint name is different.
- **Feedback Sending**: I've fixed the bug where Artemis sends feedback to Athena when saving.
- **400 OK console error**: This error is related to the text assessment events feature being generally broken (also on develop). You don't usually see the errors on develop because you usually don't turn on the Automatic Text Exercise assessment, and the text assessments are only enabled when this option is enabled. I would prefer to fix or remove that feature in a future PR.
- **Feedback Block not reusable**: Thanks, I'll look into that.

MaximilianAnzinger reviewed (dismissed):
Tested on Test Server. Everything works as described. Couldn't find any unexpected errors 👍

TheZoker reviewed (dismissed):
Tested on the provided test server and performed all the tests listed within the testing paragraph (except for the exam one). Everything worked like it was expected within the description without any additional errors! LGTM 👍

FelixTJDietrich reviewed (approved):
Code changes look good to me, I just have one comment about a potential follow-up PR. I tested the functionality on the provided test server and everything worked and appears in the logs as expected. Really great work getting Athena running again 🎉

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/TextSubmissionService.java:
```diff
@@ -139,11 +128,24 @@ public Optional<TextSubmission> getRandomTextSubmissionEligibleForNewAssessment(
+            if (athenaSubmissionId.isPresent()) {
+                var submission = textSubmissionRepository.findWithEagerResultsAndFeedbackAndTextBlocksById(athenaSubmissionId.get());
+                // Test again if it is still assessable (Athena might have taken some time to respond and another assessment might have started in the meantime):
+                if (submission.isPresent() && (submission.get().getLatestResult() == null || !submission.get().getLatestResult().isManual())) {
```
Optional: Add a `isAssessable()` method (which might be useful in other places as well)

pal03377 replied:
I'd prefer to keep it like this for now, as otherwise, I would also have to rewrite `getAssessableSubmissions` for consistency, but this method uses a lot of direct filtering in the database, and I would be unsure whether the performance implications would be too great. This seems like too much refactoring to be included in this PR, so I'll keep it like this for now.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaConnector.java:
```diff
@@ -0,0 +1,104 @@
+            response = restTemplate.postForEntity(url, httpRequestEntity, genericResponseType);
+        }
+        catch (ResourceAccessException e) {
+            log.error("Athena did not respond successfully in time: {}, stack trace: {}", e.getMessage(), e.getStackTrace());
```
You can just add the exception without any placeholders, the whole stack trace gets logged automatically.
```suggestion
            log.error("Athena did not respond successfully in time", e);
```

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaConnector.java:
```diff
@@ -0,0 +1,104 @@
+        if (responseBody == null) {
+            throw new NetworkingException("An Error occurred while calling Athena (response is null).");
+        }
+        log.debug("Finished remote call in {}ms", System.currentTimeMillis() - start);
```
Use nanoTime and TimeLogUtils.formatDuration.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaFeedbackSuggestionsService.java:
```diff
@@ -0,0 +1,103 @@
+            }).toList();
+        }
+        catch (NetworkingException error) {
+            log.error("Error while calling Athena: {}", error.getMessage());
```
```suggestion
            log.error("Error while calling Athena", error);
```

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaHealthIndicator.java:
```diff
@@ -0,0 +1,107 @@
+@Profile("athena")
+public class AthenaHealthIndicator implements HealthIndicator {
+
+    private static final String GREEN_CIRCLE = "\uD83D\uDFE2"; // unicode green circle
+
+    private static final String RED_CIRCLE = "\uD83D\uDD34"; // unicode red circle
```
```suggestion
    private static final String GREEN_CIRCLE = "\uD83D\uDFE2"; // unicode green circle 🟢

    private static final String RED_CIRCLE = "\uD83D\uDD34"; // unicode red circle 🔴
```

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSendingService.java:
```diff
@@ -0,0 +1,133 @@
+        int page = 0;
+        while (true) {
+            Pageable pageRequest = PageRequest.of(page, SUBMISSIONS_PER_REQUEST);
+            Page<TextSubmission> textSubmissions = textSubmissionRepository.findByParticipation_ExerciseIdAndSubmittedIsTrue(exercise.getId(), pageRequest);
```
You can use `nextPageable()` so you don't need to keep track of the index by yourself

pal03377 replied:
I'm now using `pageRequest.next` 👍

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaFeedbackSuggestionsService.java:
```diff
@@ -0,0 +1,103 @@
+
+        public TextExerciseDTO exercise;
+
+        public TextSubmissionDTO submission;
```
Good catch! Let's add this as an improvement later 👍

Strohgelaender commented on src/test/java/de/tum/in/www1/artemis/assessment/TutorEffortIntegrationTest.java:
```diff
@@ -69,7 +69,7 @@ void initTestCase() {
         exercise = course.getExercises().iterator().next();
         studentParticipation = studentParticipationRepository.findByExerciseId(exercise.getId()).stream().iterator().next();
-        textSubmission = textSubmissionRepository.findByParticipation_ExerciseIdAndSubmittedIsTrue(exercise.getId()).get(0);
+        textSubmission = textSubmissionRepository.findByParticipation_ExerciseIdAndSubmittedIsTrue(exercise.getId()).stream().iterator().next();
```
(assuming this returns a Set)
```suggestion
        textSubmission = textSubmissionRepository.findByParticipation_ExerciseIdAndSubmittedIsTrue(exercise.getId()).iterator().next();
```

Strohgelaender commented on src/test/java/de/tum/in/www1/artemis/AbstractAthenaTest.java:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * Create an example text exercise with feedback suggestions enabled
+     */
+    protected TextExercise createTextExercise() {
```
This should probably be located in one of the Factories?

Strohgelaender commented on src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts:
```diff
@@ -227,6 +238,88 @@ export class TextSubmissionAssessmentComponent extends TextAssessmentBaseCompone
+            }),
+        );
+        // actually consume data
+        this.feedbackSuggestionsObservable.subscribe();
```
You save this observable but never destroy it. Afaik (and I'm absolutely no expert on this, please correct me if I'm wrong) we need to manually unsubscribe from these observables when the component gets destroyed, this could otherwise lead to memory leaks.

pal03377 replied:
I'm also unsure if this would have any implications, but to be sure, I now destroy the subscription onDestroy.

Strohgelaender commented on src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts:
```diff
@@ -227,6 +238,88 @@ export class TextSubmissionAssessmentComponent extends TextAssessmentBaseCompone
+     * If the split conflicts with a manual feedback, we don't add the TextBlockRef at all.
+     *
+     * @param refToAdd The TextBlockRef to add (text block + feedback on it)
+     * @private
```
https://typedoc.org/tags/private/
"This tag should generally not be used and may be removed in a future release. The @private tag overrides the visibility of a reflection to be private."
-> has no effect since the method is already private.
```suggestion
```

Strohgelaender commented on src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts:
```diff
@@ -82,6 +86,10 @@ export class TextSubmissionAssessmentComponent extends TextAssessmentBaseCompone
     exerciseDashboardLink: string[];
     isExamMode = false;
 
+    // An observable that can be used to cancel the request for feedback suggestions
+    // if it is not done when the first manual feedback is entered
+    private feedbackSuggestionsObservable: Observable<void>;
```
This described cancelation seems to not happen anywhere.

pal03377 replied:
Now it does - only on destroy. For the case where manual feedback is added before the response completes, I decided on a different approach (not adding the automatic feedback at all). This is because HTTP fetches cannot be canceled in JS anyway.

Strohgelaender commented on src/main/webapp/app/assessment/athena.service.ts:
```diff
@@ -0,0 +1,37 @@
+export class AthenaService {
+    public resourceUrl = 'api/athena';
+
+    constructor(protected http: HttpClient, private profileService: ProfileService) {}
```
codacy issue, see also the changes in #7004.

pal03377 replied:
I tried doing that, but the ESLint pre-commit hook always seems to automatically undo my change. I'll try changing it in the next commit again.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/web/rest/TextAssessmentResource.java:
```diff
@@ -473,4 +484,17 @@ private void saveTextBlocks(final Set<TextBlock> textBlocks, final TextSubmissio
+        if (athenaFeedbackSendingService.isPresent() && exercise.isFeedbackSuggestionsEnabled()) {
+            // Athena needs IDs on all feedbacks
+            if (feedbacks.stream().anyMatch(feedback -> feedback.getId() == null)) {
+                throw new IllegalArgumentException("All feedbacks need to have an ID");
```
(Can this even happen? I assume not really since this is called directly after saving the feedback).

I don't like throwing an exception here, since this will make the request to save the feedback fail, even though the tutor feedback was saved and only an error related to Athena happened. Maybe just log a warning and don't seen the feedback in this case?

pal03377 replied:
I don't think this can ever happen, so I now removed this. If there are any problems related to this, we will notice them using the Athena logs.

Strohgelaender commented on src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts:
```diff
@@ -82,6 +86,8 @@ export class TextSubmissionAssessmentComponent extends TextAssessmentBaseCompone
     exerciseDashboardLink: string[];
     isExamMode = false;
 
+    private feedbackSuggestionsObservable: Subscription;
```
```suggestion
    private feedbackSuggestionsObservable?: Subscription;
```

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/web/rest/AthenaResource.java:
```diff
@@ -0,0 +1,76 @@
+
+        final var exercise = textExerciseRepository.findByIdElseThrow(exerciseId);
+        authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.TEACHING_ASSISTANT, exercise, null);
+        final var submission = textSubmissionRepository.findByIdElseThrow(submissionId);
```
Add a check if the passed submissionId also is part of the passed exercsie.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/scheduled/AthenaScheduleService.java:
```diff
@@ -0,0 +1,112 @@
+            return;
+        }
+        final List<TextExercise> runningTextExercises = textExerciseRepository.findAllAutomaticAssessmentTextExercisesWithFutureDueDate();
+        runningTextExercises.forEach(this::scheduleExerciseForAthena);
```
I know that we will only get SEMI-AUTOMATIC exercises from the database, however since these are two different values I'd still check is suggestions are actually enabled.
```suggestion
        runningTextExercises.forEach(this::scheduleExerciseForAthenaIfRequired);
```

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/scheduled/AthenaScheduleService.java:
```diff
@@ -0,0 +1,112 @@
+            cancelScheduledAthena(exercise.getId());
+            return;
+        }
+        // ToDo Needs to be adapted for exam exercises (@Jan Philip Bernius)
+        if (exercise.getDueDate() == null || exercise.getDueDate().compareTo(ZonedDateTime.now()) < 0) {
```
This should work imo, please double check. Also check how exam exercises are treated in the db call.
```suggestion
        if (exercise.getDueDate() == null || exerciseDateService.isAfterLatestDueDate(exercise)) {
```

In gernal, please think again about the difference between the exercise due date and the latest indivual due date. At which points should the submissions be sent?

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/web/rest/AthenaResource.java:
```diff
@@ -65,6 +65,10 @@ public ResponseEntity<List<TextBlockRef>> getFeedbackSuggestions(@PathVariable l
         final var submission = textSubmissionRepository.findByIdElseThrow(submissionId);
+        if (submission.getParticipation().getExercise().getId() != exerciseId) {
+            log.error("Exercise id {} does not match submission's exercise id {}", exerciseId, submission.getParticipation().getExercise().getId());
+            return ResponseEntity.badRequest().build();
```
We usually throw a ConflictException in this case.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/service/TextSubmissionService.java:
```diff
@@ -139,11 +128,24 @@ public Optional<TextSubmission> getRandomTextSubmissionEligibleForNewAssessment(
         }
-        var submissionWithoutResult = super.getRandomAssessableSubmission(textExercise, examMode, correctionRound);
+
+        var submissionWithoutResult = super.getRandomAssessableSubmission(assessableSubmissions);
```
I appreciate your effort in reducing database calls, however in this case we should probably reload the assessableSubmissions again to avoid randomly selecting a submission that in the meantime got locked by another tutor.

pal03377 replied:
Makes sense 👍

Strohgelaender reviewed (dismissed):
Thanks for implementing all my feedback.

tobias-lippert reviewed (approved):
Code looks good to me now. Just one question.

pal03377 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaHealthIndicator.java:
```diff
@@ -0,0 +1,107 @@
+            }
+            health.setAdditionalInfo(additionalInfo);
+        }
+        catch (Exception emAll) {
```
TBH, I also don't know where this name comes from. I just took it from the old [AtheneHealthIndicator.java](https://github.com/ls1intum/Artemis/blob/a75bf2ac7146fc8d23327838aea5d6461fd12d97/src/main/java/de/tum/in/www1/artemis/service/connectors/athene/AtheneHealthIndicator.java#L41C24-L41C24).

tobias-lippert reviewed (dismissed):
Reapprove, only exception renaming.

FelixTJDietrich reviewed (dismissed):
Code changes LGTM

Strohgelaender reviewed (dismissed):
code looks good

MaximilianAnzinger reviewed (approved):
Tested on TS3 during testing session.

FelixTJDietrich reviewed (approved):
code still looks good, only changed one line

Strohgelaender reviewed (approved):
Tested in testing session, all comments got addressed.

terlan98 reviewed (approved):
Tested on ma-schwind test server. Everything works as expected 🎉 
Great job! 🚀 

#### ✅ Submission Sending
Works as expected. The previous issue with 2 log messages is solved.

#### ✅ Submission Selection
Works as expected. 

#### ✅ ❔ Feedback Suggestions
Works as expected.

> I have a question: As far as I understand, the client does not distinguish between fully-overlapping automatic text blocks. When it receives several text blocks covering the same range, it simply replaces such blocks, which usually leads to the last overlapping block being preferred over the others. I feel like we should somehow have a way of ranking such blocks but this is definitely something for a future PR 🙂 

#### ✅ Feedback Sending
Works as expected. The feedbacks are sent only when submitting the assessment.

#### ✅ Health
Works as expected.

#### ✅ Exam Mode
Works as expected. Automatic feedbacks can't be enabled.

#### Other tests to break stuff
- ✅ 'Assess Next' button 
- ✅ Opening finished assessments
- ✅ Deleting a suggested feedback and trying to add a manual feedback for the same text block  
- ✅ Making changes to the feedback text of a suggested feedback

tobias-lippert reviewed (dismissed):
Recent code changes look good.

tobias-lippert reviewed (dismissed):
Code LGTM

tobias-lippert reviewed (dismissed):
Latest timeout code changes LGTM

FelixTJDietrich reviewed (dismissed):
Timeout code changes look good to me

tobias-lippert reviewed (approved):
Latest changes do not affect production code and look good.
Reapprove.

maximiliansoelch reviewed (approved):
Tested on the dedicated Testserver, worked as expected.
Code LGTM! 👍

FelixTJDietrich reviewed (approved):
Latest code changes also LGTM

pal03377 commented:
I changed the title to be in the `Development` category because a previously existing feature is re-implemented.
